SIPS/sip-20.md

```diff
@@ -0,0 +1,155 @@
+
+The purpose of the initial version of this SIP is to provide support for WebSockets.
+This SIP only proposes one-way communication.
+This SIP is made with extensibility in mind and can be updated to handle more external data sources (like blockchain
+events, etc.).
+
+## Motivation
+
+Snaps are currently isolated from outside events and rely on combination of user actions or CRON jobs and fetch
+endowments in order to receive extra information from independent remote sources distributed over the internet.
+
+This proposal provides an improvement that enables one way communication between a Snap and an external data or event
+source.
+The idea is that the external event source sends arbitrary data to the specific controller with listener running
+inside a wallet. The controller then sends the data to the Snap which runs execution of this feature specific RPC
+request handler.
+
+## Specification
+
+> Formal specifications are written in Typescript.
+
+### Language
+
+The key words "MUST", "MUST NOT", "SHOULD", "RECOMMENDED" and "MAY" written in uppercase in this document are to be
+interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Snap Manifest
+
+This SIP specifies a permission named `endowment:external-data`.
+
+The new field can be specified as follows in a `snap.manifest.json` file:
+
+```json
+{
+  "initialPermissions": {
+    "endowment:external-data": [
+      {
+        "type": "websocket",
+        "url": "wss://example.com/binary/endpoint",
+        "dataType": "binary"
+      },
+      {
+        "type": "websocket",
+        "url": "wss://example.com/text/endpoint",
+        "dataType": "text"
+      },
+      {
+        "type": "blockchain",
+        "chainId": "eip155:1",
+        "events": [
+          "block"
+        ]
+      }
+    ]
+  }
+}
+```
+
+### Permission caveats
+
+Each caveat object MUST have `type` property that SHOULD be `websocket` or `blockchain`.
+If the `type` is `websocket`then the caveat MUST have `url` and `dataType` properties.
+The `url` MUST be string.
+The `dataType` MUST be string that MUST be `"binary"` or `"text"`.
+The `chainId` MUST be string that represents supported blockchain.
+The `events` SHOULD be an array of strings which represent name of the events for subscription.
+
+- `type` property represents the type of data source.
+- `url` is the WebSocket URL.
+- `dataType` is type of data that WebSocket returns.
+- `chainId` is the blockchain ID.
+- `events` is the array of events to subscribe.
+
+The caveats MUST NOT have duplicate objects with the same `url` or `chainId` properties.
+
+The `url` MUST start with `wss://` which is a protocol indicator for secure WebSocket connection.
+
+### Snap implementation (request handler)
+
+This SIP proposes the following RPC request handler to execute implemented functionality for external data events:
```

Comment:
Again, I think we should start with a description of the new thing.

```suggestion
This SIP introduces a new `onExternalData` function export that MAY be
implemented by the Snap. The function is called every time the WebSocket client
receives some data. The function SHOULD accept an object parameter that MUST
have `data`, `type` and `source` properties.

The parameters represent the following:

- `data`: An object that represents the data that was received. Depending on
  the `data.type`, this SHOULD contain a `message` property, which SHOULD be
  either a `string` or an `ArrayBuffer`.
- `type`: The type of the external data source. For this SIP, this SHOULD always
  be `websocket`.
- `source`: The URL that the message was sent from. Snaps can use this to
  differentiate between different endpoints. This SHOULD be the exact same URL
  as specified in manifest.

The specification of types of `onExternalData` and its parameters:

```typescript
/**
 * A text message received from a WebSocket.
 *
 * @property type - The type of the message.
 * @property message - The message as a string.
 */
type WebSocketTextMessage = {
  type: 'text';
  message: string;
};

/**
 * A binary message received from a WebSocket.
 *
 * @property type - The type of the message.
 * @property message - The message as an ArrayBuffer.
 */
type WebSocketBinaryMessage = {
  type: 'binary';
  message: ArrayBuffer;
};

/**
 * A message received from a WebSocket.
 *
 * @property type - The type of the message.
 * @property message - The message as a string or an ArrayBuffer, depending on
 * the type.
 */
type WebSocketData = {
  type: 'websocket';
  data: WebSocketTextMessage | WebSocketBinaryMessage;
};

/**
 * The data received from an external source.
 *
 * @property type - The type of the external data source.
 * @property data - The data received from the external data source.
 */
type ExternalData = WebSocketData;

type OnExternalDataHandlerArgs = ExternalData & {
  source: string;
};

type OnExternalDataHandler = (args: OnExternalDataHandlerArgs) => Promise<void>;
```

Reply:
Updated following your suggestion.